runtime-modules/content-directory/src/lib.rs

```diff
@@ -1643,6 +1825,98 @@ impl<T: Trait> Module<T> {
         }
     }
 
+    /// Ensure property value hash with `unique` flag set is `unique` on `Class` level
+    pub fn ensure_property_value_hash_unique_option_satisfied(
+        class_id: T::ClassId,
+        property_id: PropertyId,
+        unique_property_value_hash: &T::Hash,
+    ) -> Result<(), &'static str> {
+        ensure!(
+            !<UniquePropertyValueHashes<T>>::exists(
+                (class_id, property_id),
+                unique_property_value_hash
+            ),
+            ERROR_PROPERTY_VALUE_SHOULD_BE_UNIQUE
+        );
+        Ok(())
+    }
+
+    /// Ensure all property value hashes with `unique` flag set are `unique` on `Class` level
+    pub fn ensure_property_value_hashes_unique_option_satisfied(
+        class_id: T::ClassId,
+        unique_property_value_hashes: &BTreeMap<PropertyId, T::Hash>,
+    ) -> Result<(), &'static str> {
+        for (&property_id, unique_property_value_hash) in unique_property_value_hashes {
+            Self::ensure_property_value_hash_unique_option_satisfied(
+                class_id,
+                property_id,
+                unique_property_value_hash,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// Compute old and new vec unique property value hash.
+    /// Ensure new property value hash with `unique` flag set is `unique` on `Class` level
+    pub fn ensure_vec_property_value_hashes(
+        class_id: T::ClassId,
+        in_class_schema_property_id: PropertyId,
+        property_value_vector_updated: &StoredPropertyValue<T>,
+        property_value_vector: VecStoredPropertyValue<T>,
+    ) -> Result<(T::Hash, T::Hash), &'static str> {
+        // Compute new hash from unique property value and its respective property id
+        let new_property_value_hash =
+            property_value_vector_updated.compute_unique_hash(in_class_schema_property_id);
+
+        // Ensure `Property` with `unique` flag set is `unique` on `Class` level
+        Self::ensure_property_value_hash_unique_option_satisfied(
+            class_id,
+            in_class_schema_property_id,
+            &new_property_value_hash,
+        )?;
+
+        // Compute old hash from the old unique property value and its respective property id
+        let old_property_value_hash =
+            property_value_vector.compute_unique_hash(in_class_schema_property_id);
+
+        Ok((new_property_value_hash, old_property_value_hash))
+    }
+
+    /// Compute old and new unique property value hashes.
+    /// Ensure new property value hashes with `unique` flag set are `unique` on `Class` level
+    pub fn ensure_property_values_hashes(
```

Comment:
This ensure checkers has a name that does not really indicate a clear purpose.

Closer inspection shows that it seems to do three things
1) check that new property values respect uniquness constraint
2) if so, compute hashes representing these values values
3) if so, determine what hashes must be removed from the uniquness index because they represent values that will no longer be in directory.

It seems that only 1 or 2 are really part of the ensure check, and you could perhaps have one checker with a name like `ensure_new_property_values_respect_uniquness`. It can then return hashes in success case. Then for 3, which is only relevant if you go to mutation, you do the call to `compute_old_unique_hashes` somewhere else. Could be a standalone call, or it could be inside `remove_unique_property_value_hashes` or something else.

Reply:
Agree, refactored it here https://github.com/Joystream/joystream/pull/1064/commits/673fcb38407d012598e14158867d090e6a7846aa
Thank you for the good explanation!